tutorials/haskell/snap-web-app/tutorial.md

```diff
@@ -0,0 +1,482 @@
+that includes the state of all snaplets used by our application in a module
+called [Application](https://github.com/stackbuilders/tutorials/tutorials/haskell/snap-web-app/code/src/Todo/Application.hs).
+
+```haskell
+data App = App
+  { _heist :: Snaplet (Heist App)
+  , _conn  :: Connection
+  }
+
+```
+This data structure contains a `_heist` field holding the
+[Heist Snaplet](http://snapforbeginners.com/chapters/heist.html)  and a `_conn`
+for the database connection. This is the right place to include all the Snaplets you
+want if you're aiming to build a bigger app. Also, this data structure will be used
+later to initialize the web app in the
+[Site module](https://github.com/stackbuilders/tutorials/tutorials/haskell/snap-web-app/code/src/Todo/Site.hs#L139-L146)
+like follows:
+
+```haskell
+app :: SnapletInit App App
+app = makeSnaplet "app" "A todo application." Nothing $ do
+  h <- nestSnaplet "heist" heist $ heistInit "templates"
+  c <- liftIO $ Db.startConnection "todos.db"
+  liftIO $ Db.checkMigrationOrMigrate c
+
+  addRoutes routes
+  return $ App h c
+```
+All snaplet initializers must be wrapped in a call to [`makeSnaplet`](http://hackage.haskell.org/package/snap-0.14.0.6/docs/Snap-Snaplet.html#v:makeSnaplet),
+that's why this `app` function uses it.
+If you take a closer look, the initializer computation starts up Heist and the database connection.
+It also configure routes and checks for migrations, but we'll get to that later.
+So, this module is really important as it holds our application stack and resource handlers.
+
+As for the [Main module](https://github.com/stackbuilders/tutorials/tutorials/haskell/snap-web-app/code/src/Todo/Main.hs),
+it just bootstraps the web server using the previous Snaplet created in the
+function `app` and the application configuration. You rarely will need to modify
+it, but we encourage you take a look at it and understand what happens there.
+
+## Setting up the model and migrations
+
+This application goes about managing Todo items, which requires a very simple
+data type definition with fields for description, completion status and,
+of course, a primary key. This is defined in
+[Db module](https://github.com/stackbuilders/tutorials/tutorials/haskell/snap-web-app/code/src/Todo/Db.hs)
+and this is how it looks like using
+[record syntax](https://en.wikibooks.org/wiki/Haskell/More_on_datatypes#Named_Fields_.28Record_Syntax.29):
+
+```haskell
+data Todo = Todo
+  { todoID      :: Maybe Int
+  , description :: Text
+  , completed   :: Bool
+  } deriving (Eq, Show, Read)
+```
+[Back in the Site module](https://github.com/stackbuilders/tutorials/tutorials/haskell/snap-web-app/code/src/Todo/Site.hs#L143),
+there were used a function called `checkMigrationOrMigrate`
+receiving a database connection as parameter, while initalizing our Todo App.
+This function verifies if `todos` table exists in the database and if not, then creates it.
```

Comment:
@nebtrx In general, use colons instead of periods before code ("then creates it**:**).

I don't understand the "there were used a function..." part. Is it "Back in the Site module, we used a function called..."?